### PR TITLE
Builtin tls helper

### PR DESCRIPTION
**Pull request by i0rek**
This PR adds builtin helpers to setup agent CAs and certificates for Consul TLS!

### CODE REVIEW

File: website/source/docs/guides/creating-certificates.html.md [hunk 1]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+and client certificates allows tools like `curl` to be able to communicate with
+Consul's HTTPS API when run on the same host. Other SANs may be added during
+server/client certificates creation with `-additional-dnsname` to allow remote
+HTTPS requests from other hosts.
+
+## Configuring the Consul UI for HTTPS
+
+If your servers and clients are configured now like above, you won't be able to
+access the builtin UI anymore. We recommend that you pick a Consul agent you
+want to run the UI on and follow the instructions to get the UI up and running
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: Maybe "pick one (or two for availability)"

File: website/source/docs/guides/creating-certificates.html.md [hunk 2]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+again.
+
+### Step 1: Which interface to bind to?
+
+Depending on your setup you might need to change to which interface you are
+binding because thats `127.0.0.1` by default for the UI. Either via the
+[`addresses.https`](https://www.consul.io/docs/agent/options.html#https) or
+[client_addr](https://www.consul.io/docs/agent/options.html#client_addr) option
+which also impacts the DNS server. The Consul UI is unproteced which means you
+need to put some auth in front of it if you want to make it publicly available!
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: Should we add a note here about `enable_script_checks` and `remote exec`. They MUST NOT be enabled on these UI agents that expose HTTP to the network?

File: website/source/docs/guides/creating-certificates.html.md [hunk 3]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+
+```json
+{
+  "verify_incoming": false,
+  "verify_incoming_rpc": true
+}
+```
+
+~> This is the only time we are changing the value of the existing option
+`verify_incoming` to false. Make sure to change it in the correct place!
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: Maybe reword this to clarify that the "correct place" means, only do this on UI servers?

File: website/source/docs/guides/creating-certificates.html.md [hunk 4]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+$ curl https://localhost:8501/ui/ -k -I
+HTTP/2 200
+...
+```
+
+### Step 3: Subject Alternative Name
+
+This step will take care of setting up the domain you want to use to access the
+Consul UI. Unless thats `localhost` or `127.0.0.1` you will have to go through
+that because it won't work otherwise:
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: This sentence confused me, do you mean:

> Unless you only need to access the UI over `localhost` or `127.0.0.1` you will need to go complete this step.

File: website/source/docs/guides/creating-certificates.html.md [hunk 5]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+
+```shell
+$ curl https://myconsului.com:8501/ui/ \
+  --resolve 'myconsului.com:8501:127.0.0.1' \
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: nit: maybe `consul.example.com` just incase someone buys that domain and it is mistaken for an official thing?

File: website/source/docs/guides/creating-certificates.html.md [hunk 6]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+  --cacert consul-ca.pem
+curl: (51) SSL: no alternative certificate subject name matches target host name 'myconsului.com'
+...
+```
+
+The above command simulates a request a browser is making when you are trying to
+use the domain `myconsului.com` to access your UI. This time we do validate the
+certificates because the browser is doing the same and we also pass the Consul
+CA. The problem this time is that your domain is not in `Subject Alternative
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: > This time we do validate the certificates because the browser is doing the same and we also pass the Consul CA.

Not sure what that is trying to say. IMO this paragraph reads fine without that sentence.

File: website/source/docs/guides/creating-certificates.html.md [hunk 7]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+  --cacert consul-ca.pem -I
+HTTP/2 200
+...
+```
+
+### Step 4: Trust the Consul CA
+
+So far we have provided curl with our CA so that it can verify the connection,
+but if we stop doing that it will complain and so will our browser if you are
+going to your UI on https://myconsului.com:
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: `s/are going to/visit/`

File: website/source/docs/guides/creating-certificates.html.md [hunk 8]
```diff
@@ -192,7 +267,149 @@ $ export CONSUL_CLIENT_KEY=cli-key.pem
...
+  --resolve 'myconsului.com:8501:127.0.0.1' -I
+HTTP/2 200
+...
+```
+
+## Summary
+
+When you completed this guide, your Consul cluster has TLS enabled and is
+protected against all sorts of attacks plus you made sure only your servers are
+able to start as a server! A good next step would be to read about [ACLs][acl]!
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: Suggest:

> When you have completed this guide, your Consul cluster will have TLS enabled and will encrypt all RPC and HTTP traffic (assuming you disabled the [HTTP port](#link-to-step)). The other pre-requisites for a secure Consul deployment are:
  - [Enable gossip encryption](...)
  - [Configure ACLs][acl] with default deny

File: command/tls/ca/tls_ca.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  This command has subcommands for interacting with Certificate Authorities.
+
+  Here are some simple examples, and more detailed examples are available
+  in the subcommands or the documentation.
+
+  Create a CA
+
+    $ consul tls ca create
+    ==> saved consul-ca.pem
+    ==> saved consul-ca-key.pem
```
[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/ca/tls_ca.go]: I would like to remove the prefix for CAs and Certs entirely. The ca would be called `agent-ca.pem`. That better?

File: command/tls/generate.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	var buf bytes.Buffer
+	err = pem.Encode(&buf, &pem.Block{Type: "EC PRIVATE KEY", Bytes: bs})
+	if err != nil {
+		return nil, "", fmt.Errorf("error encoding private key: %s", err)
+	}
+
+	return pk, buf.String(), nil
+}
+
+// GenerateCA generates a new CA
```
[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/generate.go]: Done.

File: website/source/docs/guides/creating-certificates.html.md [hunk 11]
```diff
@@ -1,20 +1,53 @@
 ---
 layout: "docs"
-page_title: "Creating Certificates"
+page_title: "Creating and Configuring Certificates"
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: Sure!

File: website/source/docs/guides/creating-certificates.html.md [hunk 12]
```diff
@@ -1,20 +1,53 @@
...
+
+## Creating Certificates
+
+### Estimated Time to Complete
+
+2 minutes
+
+### Prerequisites
+
+This guide assumes you have Consul 1.4 (or newer) in your PATH.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 13]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
 ~> Consul certificates may be signed by intermediate CAs as long as the root CA
    is the same. Append all intermediate CAs to the `cert_file`.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: Good point! Given how this guide changed to a step by step thing I think we could remove this note. It is a guide for a simple setup with tooling support - intermediates are outside of the scope.

File: website/source/docs/guides/creating-certificates.html.md [hunk 14]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
 
-- [Encryption](/docs/agent/encryption.html)
+```shell
+$ consul tls ca create
+==> Saved consul-ca.pem
+==> Saved consul-ca-key.pem
+```
 
-## Estimated Time to Complete
+The CA certificate (`consul-ca.pem`) contains the public key necessary to
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 15]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+validate Consul certificates and therefore must be distributed to every node
+that requires access.
 
-20 minutes
+~> The CA key (`consul-ca-key.pem`) will be used to sign certificates for Consul
+nodes and must be kept private.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 16]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+    and all ACL tokens. Do not distribute them to production hosts
+    that are not server nodes. Store them as securely as CA keys.
+==> Using consul-ca.pem and consul-ca-key.pem
+==> Saved consul-server-dc1-0.pem
+==> Saved consul-server-dc1-0-key.pem
+```
 
-### Step 1: Create Certificate Authority
+Please repeat this process until there is an *individual* certificate for each
+server. The command can be called over and over again, it will automatically add
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: I had similar thoughts. Since it is mentioned in our security model I thought I add that to our guide as well.

> . At a minimum, verify_outgoing should be enabled to verify server authenticity with each server having a unique TLS certificate.

Plus as you said the user might be able to revoke at some point. I don't feel strong about it, but leaning towards leaving it in.

File: website/source/docs/guides/creating-certificates.html.md [hunk 17]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
-There are a variety of tools for managing your own CA, [like the PKI secret
-backend in Vault][vault-pki], but for the sake of simplicity this guide will
-use [cfssl][cfssl]. You can generate a private CA certificate and key with
-[cfssl][cfssl]:
+In order to authenticate Consul servers, servers are provided with a special
+certificate - one that contains `server.dc1.consul` in the `Subject Alternative
+Name`. If you enable
+[`verify_server_hostname`](/docs/agent/options.html#verify_server_hostname),
+only agents that provide such certificate are allowed to boot as a server.
+An attacker could otherwise compromise a Consul agent and restart the agent as a
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 18]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+server in order to get access to all the data in your cluster! This is why
+server certificates are special, and only servers should have them provisioned.
+
+~> Server certificates, like the CA key, must be kept private!
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 19]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+Client certificates are also signed by your CA, but they do not have that
+special `Subject Alternative Name` which means that if `verify_server_hostname`
+is enabled, they cannot start as a server.
+
+## Configuring Agents
+
+### Prerequisites
+
+For this section you need access to your existing or new Consul cluster and have
+the certificates from previous chapter available.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md [hunk 20]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+all into one file if you prefer that.
+
+### Introduction
+
+By now you have created the certificates you need to enable TLS in your cluster.
+The next steps will demonstrate how to configure TLS. However take this with a
+grain of salt because turning on TLS might not be that straight forward
+depending on your setup. For example in a cluster that is being used in
+production you want to turn on TLS step by step to ensure there is no downtime
+like described [here][guide].
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: I like your version better!

File: website/source/docs/guides/creating-certificates.html.md [hunk 21]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+  "cert_file": "consul-server-dc1-0.pem",
+  "key_file": "consul-server-dc1-0-key.pem"
 }
 ```
 
-```shell
-# Generate the CA's private key and certificate
-$ cfssl gencert -initca ca-csr.json | cfssljson -bare consul-ca
-```
+After a Consul agent restart, your servers should be only talking TLS.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: A client can still connect via HTTP - `verify_incoming` only works on a TLS connection.

I agree that we should suggest disabling the HTTP port as well. I will put a note there about the implications for the clients.

[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]:
```json
{
  "verify_incoming": true,
  "verify_outgoing": true,
  "verify_server_hostname": true,
  "ca_file": "agent-consul-ca.pem",
  "cert_file": "server-dc1-consul-0.pem",
  "key_file": "server-dc1-consul-0-key.pem",
  "ports": {
    "http": -1,
    "https": 8501
  }
}
```

This configuration disables the HTTP port to make sure there is only encryted
communication. Existing agents that are not yet prepared to talk HTTPS won't be
able to connect afterwards.

[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: Now it is actually true that there is only TLS :).

File: website/source/docs/guides/creating-certificates.html.md [hunk 22]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+  "verify_incoming": true,
+  "verify_outgoing": true,
+  "verify_server_hostname": true,
+  "ca_file": "consul-ca.pem",
+  "cert_file": "consul-client-0.pem",
+  "key_file": "consul-client-0-key.pem"
+}
+```
+
+After a Consul agent restart, your agents should be only talking TLS.
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: I mode the same changes as above here as well.

File: website/source/docs/guides/creating-certificates.html.md [hunk 23]
```diff
@@ -25,159 +58,201 @@ signed by this CA will be allowed to communicate with the cluster.
...
+    "ports": {
+        "http": -1,
+        "https": 8501
     }
-  }
 }
 ```
 
+If your cluster is configured like that, you will need to create additional
+certificates in order to be able to continue to access the API and the UI:
```
[the remark this answers is not in the document]
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: I changed the config to disable HTTP and now we are good?

File: website/source/docs/guides/creating-certificates.html.md (hunk 1 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md (hunk 2 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: Only HTTPS is exposed at this point. Do you think we should still mention it?

Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ~> Since your Consul agent is now available to the network,
`enable_script_checks` and `remote exec` must be disabled!

File: website/source/docs/guides/creating-certificates.html.md (hunk 3 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md (hunk 4 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: Went with your suggestion.

File: website/source/docs/guides/creating-certificates.html.md (hunk 5 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: good catch!

File: website/source/docs/guides/creating-certificates.html.md (hunk 6 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: website/source/docs/guides/creating-certificates.html.md (hunk 7 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

**Comment by i0rek** [comment]
I updated https://tlsdocs.netlify.com/docs/guides/creating-certificates.html.

File: website/source/docs/guides/creating-certificates.html.md (hunk 8 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: command/tls/ca/tls_ca.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (banks) [reply command/tls/ca/tls_ca.go]: Maybe it's too long but I think a reasonable default is `consul-agent-ca.pem` that way when they loo in a dir full of certs later it's super clear what that one is for etc.

[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/ca/tls_ca.go]: I made a mistake, I wanted to write `agent-consul-ca.pem`. Consul stands for the domain since the domain is important with `-name-constraint`.

File: website/source/docs/guides/creating-certificates.html.md [hunk 24]
```diff
@@ -296,10 +305,15 @@ Binding to `0.0.0.0` should work:
...
   "ui": true,
-  "client_addr": "0.0.0.0"
+  "client_addr": "0.0.0.0",
+  "enable_script_checks": false,
+  "disable_remote_exec": true
 }
 ```
 
+~> Since your Consul agent is now available to the network,
+`enable_script_checks` and `remote exec` must be disabled!
```
Comment (banks) [inline website/source/docs/guides/creating-certificates.html.md]: nit: maybe link to the options in docs and be consistent so use `[disable_remote_exec](https://www.consul.io/docs/agent/options.html#disable_remote_exec)` (i,e. the config param rather than the command name? It's a little awkward that one is enable and the other disable but still...


> ~> Since your Consul agent's HTTPS API is available to the network, please make sure [`enable_script_checks`](https://www.consul.io/docs/agent/options.html#_enable_script_checks) is `false` and [`disable_remote_exec`](https://www.consul.io/docs/agent/options.html#disable_remote_exec) is `true`.

File: command/tls/ca/create/tls_ca_create.go [hunk 25]
```diff
@@ -40,8 +48,8 @@ func (c *cmd) Run(args []string) int {
 		return 1
 	}
 
-	certFileName := fmt.Sprintf("%s-agent-ca.pem", c.prefix)
-	pkFileName := fmt.Sprintf("%s-agent-ca-key.pem", c.prefix)
+	certFileName := fmt.Sprintf("agent-%s-ca.pem", c.domain)
```
[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/ca/create/tls_ca_create.go]: The reason why I removed the consul in the beginning is that it gets ridiculous for certs with the consul domain: `consul-server-dc1-consul-0.pem`. Since was gone for certs, I removed it for the CA as well. I chose the order because this is also what the SAN looks like `server-dc1-consul-0.pem` becomes `server.dc1.consul`. 
I agree that having the domain in the first for the filename place reads a little nicer, but it then no longer has the same order as the  SAN.

File: website/source/docs/guides/creating-certificates.html.md (hunk 24 above)
Comment (i0rek) [reply website/source/docs/guides/creating-certificates.html.md]: ✔️

File: command/tls/ca/create/tls_ca_create.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (banks) [reply command/tls/ca/create/tls_ca_create.go]: I'm not sure I understand.

`consul-example-com-agent-ca.pem`

seems more consistent with

`server-dc1-consul-0.pem`

than `agent-consul-example-com-ca.pem` to me?

Not a huge deal though, I'll leave it to you but maybe listing a full example of all the CA and cert filenames for both default and custom domain might help double check they aren't confusingly inconsistent anywhere?

[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/ca/create/tls_ca_create.go]: This is the current state:

| dc,domain  | ca | server | client | cli
| ---- | ---- | ---- | ---- | ---- |
| default, default | agent-consul-ca.pem  |  server-dc1-consul-0.pem |  client-dc1-consul-0.pem |  cli-dc1-consul-0.pem
| dc2, example.com | agent-example.com-ca.pem  | server-dc2-example.com-0.pem | client-dc2-example.com-0.pem | cli-dc2-example.com-0.pem

Your proposal:

| dc,domain  | ca | server | client | cli
| ---- | ---- | ---- | ---- | ---- |
| default, default | consul-consul-agent-ca.pe  |  server-dc1-consul-0.pem |  client-dc1-consul-0.pem |  cli-dc1-consul-0.pem
| dc2, example.com |consul-example-com-agent-ca.pe  | server-dc2-example.com-0.pem | client-dc2-example.com-0.pem | cli-dc2-example.com-0.pem

[the remark this answers is not in the document]
Comment (banks) [reply command/tls/ca/create/tls_ca_create.go]: Ah, I see. No my proposal was meant to be (I may have mis-stated it) "<domain>-agent-ca.pem"

So it would `consul-agent-ca.pem` by default (which reads nicely to me) but with `consul.example.com` as the domain `consul-example-com-agent-ca.pem`.

Like I said no big deal. The one thing that makes it worth nailing this down and merging soon though is that the Encryption guide @kaitlincarter-hc is working on in #5059 would really benefit from showing the same default file names that this tool creates!

[the remark this answers is not in the document]
Comment (i0rek) [reply command/tls/ca/create/tls_ca_create.go]: ok, I will change it so that the domain is first:

| dc,domain  | ca | server | client | cli
| ---- | ---- | ---- | ---- | ---- |
| default, default | consul-agent-ca.pem  |  consul-server-dc1-0.pem |  consul-client-dc1-0.pem |  consul-cli-dc1-0.pem
| dc2, example.com | example.com-agent-ca.pem  | example.com-server-dc2-0.pem | example.com-client-dc2-0.pem | example.com-cli-dc2-0.pem

[the remark this answers is not in the document]
Comment (banks) [reply command/tls/ca/create/tls_ca_create.go]: 🎉 Let's merge this thing!
